File: packages/kotlin/src/test/kotlin/com/luxo/client/TypesTest.kt
package com.luxo.client

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.json.Json

class PageTest {

    @Test
    fun `page data class properties`() {
        val page = Page(
            items = listOf("a", "b", "c"),
            total = 10,
            page = 1,
            pageSize = 3,
        )

        assertEquals(listOf("a", "b", "c"), page.items)
        assertEquals(10, page.total)
        assertEquals(1, page.page)
        assertEquals(3, page.pageSize)
    }

    @Test
    fun `page serialization round-trip`() {
        val json = """{"items":["x","y"],"total":5,"page":2,"pageSize":2}"""
        val page = Json.decodeFromString<Page<String>>(json)
        assertEquals(listOf("x", "y"), page.items)
        assertEquals(5, page.total)
        assertEquals(2, page.page)
        assertEquals(2, page.pageSize)

        val encoded = Json.encodeToString(Page.serializer(String.serializer()), page)
        assertEquals(json, encoded)
    }
}

class LuxoSchemaTest {

    @Test
    fun `schema defaults to empty maps`() {
        val schema = LuxoSchema()
        assertEquals(emptyMap(), schema.models)
        assertEquals(emptyMap(), schema.apis)
        assertEquals(emptyMap(), schema.enums)
        assertEquals(emptyMap(), schema.types)
    }

    @Test
    fun `schema deserialization`() {
        val json = """
        {
            "models": {
                "User": {
                    "name": "User",
                    "fields": [
                        {"id": 1, "name": "id", "type": "Int"},
                        {"id": 2, "name": "name", "type": "String", "nullable": true}
                    ]
                }
            },
            "enums": {
                "Role": {"name": "Role", "values": ["ADMIN", "USER"]}
            },
            "apis": {},
            "types": {}
        }
        """.trimIndent()

        val schema = Json.decodeFromString<LuxoSchema>(json)

        assertEquals(1, schema.models.size)
        val user = schema.models["User"]!!
        assertEquals("User", user.name)
        assertEquals(2, user.fields.size)
        assertEquals(1, user.fields[0].id)
        assertEquals("id", user.fields[0].name)
        assertEquals("Int", user.fields[0].type)
        assertFalse(user.fields[0].nullable)
        assertTrue(user.fields[1].nullable)

        val role = schema.enums["Role"]!!
        assertEquals("Role", role.name)
        assertEquals(listOf("ADMIN", "USER"), role.values)
    }
}

class LuxoFieldTest {

    @Test
    fun `field defaults`() {
        val field = LuxoField(id = 1, name = "test", type = "String")
        assertNull(field.typeName)
        assertFalse(field.nullable)
        assertFalse(field.isList)
        assertFalse(field.relation)
    }

    @Test
    fun `field with all properties`() {
        val field = LuxoField(
            id = 5,
            name = "posts",
            type = "Post",
            typeName = "Post",
            nullable = true,
            isList = true,
            relation = true,
        )
        assertEquals(5, field.id)
        assertEquals("posts", field.name)
        assertEquals("Post", field.type)
        assertEquals("Post", field.typeName)
        assertTrue(field.nullable)
        assertTrue(field.isList)
        assertTrue(field.relation)
    }
}

class LuxoAPITest {

    @Test
    fun `api defaults`() {
        val api = LuxoAPI(id = 1, name = "getUser", module = "user")
        assertNull(api.returnType)
        assertFalse(api.returnList)
        assertFalse(api.paginated)
        assertEquals(emptyList(), api.params)
    }

    @Test
    fun `api with params`() {
        val api = LuxoAPI(
            id = 10,
            name = "listUsers",
            module = "user",
            returnType = "User",
            returnList = true,
            paginated = true,
            params = listOf(
                LuxoParam(id = 1, name = "page", type = "Int"),
                LuxoParam(id = 2, name = "pageSize", type = "Int"),
            ),
        )
        assertEquals(10, api.id)
        assertEquals("listUsers", api.name)
        assertEquals("user", api.module)
        assertEquals("User", api.returnType)
        assertTrue(api.returnList)
        assertTrue(api.paginated)
        assertEquals(2, api.params.size)
        assertEquals("page", api.params[0].name)
    }
}

class TransportModeTest {

    @Test
    fun `enum values`() {
        assertEquals(2, TransportMode.entries.size)
        assertEquals(TransportMode.JSON, TransportMode.valueOf("JSON"))
        assertEquals(TransportMode.BINARY, TransportMode.valueOf("BINARY"))
    }
}

File: packages/dart/test/types_test.dart
import 'package:test/test.dart';
import 'package:luxo_client/src/types.dart';

void main() {
  group('Page', () {
    test('fromJson parses correctly', () {
      final json = {
        'items': [
          {'id': 1, 'name': 'Alice'},
          {'id': 2, 'name': 'Bob'},
        ],
        'total': 10,
        'page': 1,
        'pageSize': 2,
      };
      final page = Page.fromJson(json, (e) => e);
      expect(page.items.length, equals(2));
      expect(page.total, equals(10));
      expect(page.page, equals(1));
      expect(page.pageSize, equals(2));
      expect(page.items[0]['name'], equals('Alice'));
    });

    test('fromJson with empty items', () {
      final json = {
        'items': [],
        'total': 0,
        'page': 1,
        'pageSize': 10,
      };
      final page = Page.fromJson(json, (e) => e);
      expect(page.items, isEmpty);
      expect(page.total, equals(0));
    });
  });

  group('LuxoField', () {
    test('fromJson with all fields', () {
      final json = {
        'id': 1,
        'name': 'username',
        'type': 'String',
        'typeName': 'User',
        'nullable': true,
        'isList': true,
        'relation': true,
      };
      final field = LuxoField.fromJson(json);
      expect(field.id, equals(1));
      expect(field.name, equals('username'));
      expect(field.type, equals('String'));
      expect(field.typeName, equals('User'));
      expect(field.nullable, isTrue);
      expect(field.isList, isTrue);
      expect(field.relation, isTrue);
    });

    test('fromJson with defaults', () {
      final json = {
        'id': 2,
        'name': 'email',
        'type': 'String',
      };
      final field = LuxoField.fromJson(json);
      expect(field.nullable, isFalse);
      expect(field.isList, isFalse);
      expect(field.relation, isFalse);
      expect(field.typeName, isNull);
    });
  });

  group('LuxoModel', () {
    test('fromJson', () {
      final json = {
        'name': 'User',
        'fields': [
          {'id': 1, 'name': 'id', 'type': 'Int'},
          {'id': 2, 'name': 'name', 'type': 'String'},
        ],
      };
      final model = LuxoModel.fromJson(json);
      expect(model.name, equals('User'));
      expect(model.fields.length, equals(2));
      expect(model.fields[0].name, equals('id'));
    });
  });

  group('LuxoAPI', () {
    test('fromJson with all fields', () {
      final json = {
        'id': 1,
        'name': 'getUser',
        'module': 'user',
        'returnType': 'User',
        'returnList': true,
        'paginated': true,
        'params': [
          {'id': 1, 'name': 'userId', 'type': 'Int'},
        ],
      };
      final api = LuxoAPI.fromJson(json);
      expect(api.id, equals(1));
      expect(api.name, equals('getUser'));
      expect(api.module, equals('user'));
      expect(api.returnType, equals('User'));
      expect(api.returnList, isTrue);
      expect(api.paginated, isTrue);
      expect(api.params.length, equals(1));
      expect(api.params[0].name, equals('userId'));
    });

    test('fromJson with defaults', () {
      final json = {
        'id': 2,
        'name': 'listPosts',
        'module': 'post',
      };
      final api = LuxoAPI.fromJson(json);
      expect(api.returnType, isNull);
      expect(api.returnList, isFalse);
      expect(api.paginated, isFalse);
      expect(api.params, isEmpty);
    });
  });

  group('LuxoEnum', () {
    test('fromJson', () {
      final json = {
        'name': 'Status',
        'values': ['Active', 'Inactive', 'Pending'],
      };
      final e = LuxoEnum.fromJson(json);
      expect(e.name, equals('Status'));
      expect(e.values, equals(['Active', 'Inactive', 'Pending']));
    });
  });

  group('LuxoSchema', () {
    test('fromJson with all sections', () {
      final json = {
        'models': {
          'User': {
            'name': 'User',
            'fields': [
              {'id': 1, 'name': 'id', 'type': 'Int'},
            ],
          },
        },
        'apis': {
          'getUser': {
            'id': 1,
            'name': 'getUser',
            'module': 'user',
          },
        },
        'enums': {
          'Status': {
            'name': 'Status',
            'values': ['Active'],
          },
        },
        'types': {
          'Address': {
            'name': 'Address',
            'fields': [
              {'id': 1, 'name': 'street', 'type': 'String'},
            ],
          },
        },
      };
      final schema = LuxoSchema.fromJson(json);
      expect(schema.models.containsKey('User'), isTrue);
      expect(schema.apis.containsKey('getUser'), isTrue);
      expect(schema.enums.containsKey('Status'), isTrue);
      expect(schema.types.containsKey('Address'), isTrue);
    });

    test('fromJson with empty/null sections', () {
      final schema = LuxoSchema.fromJson({});
      expect(schema.models, isEmpty);
      expect(schema.apis, isEmpty);
      expect(schema.enums, isEmpty);
      expect(schema.types, isEmpty);
    });
  });

  group('LuxoParam', () {
    test('fromJson', () {
      final json = {'id': 3, 'name': 'limit', 'type': 'Int'};
      final param = LuxoParam.fromJson(json);
      expect(param.id, equals(3));
      expect(param.name, equals('limit'));
      expect(param.type, equals('Int'));
    });
  });

  group('LuxoTypeDecl', () {
    test('fromJson', () {
      final json = {
        'name': 'Address',
        'fields': [
          {'id': 1, 'name': 'city', 'type': 'String'},
          {'id': 2, 'name': 'zip', 'type': 'String'},
        ],
      };
      final t = LuxoTypeDecl.fromJson(json);
      expect(t.name, equals('Address'));
      expect(t.fields.length, equals(2));
    });
  });
}
